pkg/azure/aks/machine_store.go

```diff
@@ -0,0 +1,253 @@
+	if err != nil {
+		logger.LogAttrs(parentCtx, slog.LevelError, "unable to create compute client factory", slog.String("err", err.Error()))
+		return nil, ErrClientCreationFailure
+	}
+
+	containerClientFactory, err := armcontainerservice.NewClientFactory(subscriptionId, credentials, nil)
+	if err != nil {
+		logger.LogAttrs(parentCtx, slog.LevelError, "unable to create container client factory", slog.String("err", err.Error()))
+		return nil, ErrClientCreationFailure
+	}
+
+	ms := &MachineStore{
+		context: parentCtx,
+		logger:  logger,
+
+		azResourceGroupClient: rgClient,
+		azVMSSClient:          computeClientFactory.NewVirtualMachineScaleSetsClient(),
+		azVMSSVmClient:        computeClientFactory.NewVirtualMachineScaleSetVMsClient(),
+		azAksClient:           containerClientFactory.NewManagedClustersClient(),
+
+		MachineMap:     make(map[string]*VirtualMachineInfo),
+		machineMapLock: &sync.RWMutex{},
+	}
+
+	go func() {
+		err := ms.PopulateMachineStore(ms.context)
+		if err != nil {
+			ms.logger.LogAttrs(ms.context, slog.LevelError, "error populating initial machine store", slog.String("error", err.Error()))
+		}
+	}()
+
+	return ms, nil
+}
+
+func (m *MachineStore) getVmInfoByVmName(vmName string) (*VirtualMachineInfo, error) {
+	m.machineMapLock.RLock()
+	defer m.machineMapLock.RUnlock()
+
+	vmInfo := m.MachineMap[vmName]
+	if vmInfo == nil {
+		return nil, ErrMachineNotFound
+	}
+
+	return vmInfo, nil
+}
+
+func (m *MachineStore) getVmInfoFromVmss(ctx context.Context, rgName string, vmssName string, priority MachinePriority, osInfo MachineOperatingSystem) (map[string]*VirtualMachineInfo, error) {
+	vmInfo := make(map[string]*VirtualMachineInfo)
+
+	opts := &armcompute.VirtualMachineScaleSetVMsClientListOptions{
+		Expand: to.StringPtr("instanceView"),
+	}
+	pager := m.azVMSSVmClient.NewListPager(rgName, vmssName, opts)
+	for pager.More() {
+		nextResult, err := pager.NextPage(ctx)
+		if err != nil {
+			m.logger.LogAttrs(ctx, slog.LevelError, "unable to advance page in VMSS VM Client", slog.String("err", err.Error()))
+			return nil, ErrPageAdvanceFailure
+		}
+
+		for _, v := range nextResult.Value {
+			vmName, err := getMachineName(v)
+			if err != nil {
+				m.logger.Error(err.Error())
+				continue
+			}
+
+			vmInfo[vmName] = &VirtualMachineInfo{
+				Name:            vmName,
+				Region:          to.String(v.Location),
+				OwningVMSS:      vmssName,
+				MachineTypeSku:  to.String(v.SKU.Name),
+				Priority:        priority,
+				OperatingSystem: osInfo,
+			}
+			m.logger.LogAttrs(ctx, slog.LevelDebug, "found machine information", slog.String("machineName", vmName))
+		}
+	}
+
+	return vmInfo, nil
+}
+
+func (m *MachineStore) getVmInfoFromResourceGroup(ctx context.Context, rgName, clusterName string) (map[string]*VirtualMachineInfo, error) {
+	vmInfoMap := make(map[string]*VirtualMachineInfo)
+
+	pager := m.azVMSSClient.NewListPager(rgName, nil)
```

Comment:
Suggestion, to improve the ability to test this method and make it reusable if we need to in the future, pass in the aks client as a parameter to the method and decouple the method from MachineStore.

Reply:
same as above